jetty-server/src/main/java/org/eclipse/jetty/server/PlainOrSslConnectionFactory.java

```diff
@@ -0,0 +1,192 @@
+     * behavior (by default it closes the EndPoint) for example by writing a minimal response. </p>
+     *
+     * @param sslConnectionFactory The SslConnectionFactory to use if the first bytes are TLS
+     * @param plainProtocol        the protocol of the ConnectionFactory to use if the first bytes are not TLS, or null.
+     */
+    public PlainOrSslConnectionFactory(SslConnectionFactory sslConnectionFactory, String plainProtocol)
+    {
+        super("plain|ssl");
+        this.sslConnectionFactory = sslConnectionFactory;
+        this.plainProtocol = plainProtocol;
+    }
+
+    @Override
+    public Connection newConnection(Connector connector, EndPoint endPoint)
+    {
+        return configure(new PlainOrSslConnection(endPoint, connector), connector, endPoint);
+    }
+
+    /**
+     * @param buffer The buffer with the first bytes of the connection
+     * @return whether the bytes seem TLS bytes
+     */
+    protected boolean seemsTLS(ByteBuffer buffer)
+    {
+        int tlsFrameType = buffer.get(0) & 0xFF;
+        int tlsMajorVersion = buffer.get(1) & 0xFF;
+        return (tlsFrameType == TLS_HANDSHAKE_FRAME_TYPE || tlsFrameType == TLS_ALERT_FRAME_TYPE) && tlsMajorVersion == TLS_MAJOR_VERSION;
+    }
+
+    /**
+     * <p>Callback method invoked when {@code plainProtocol} is {@code null}
+     * and the first bytes are not TLS.</p>
+     * <p>This typically happens when a client is trying to connect to a TLS
+     * port using the {@code http} scheme (and not the {@code https} scheme).</p>
+     * <p>This method may be overridden to write back a minimal response such as:</p>
+     * <pre>
+     * HTTP/1.1 400 Bad Request
+     * Content-Length: 35
+     * Content-Type: text/plain; charset=UTF8
+     * Connection: close
+     *
+     * Plain HTTP request sent to TLS port
+     * </pre>
+     *
+     * @param buffer   The buffer with the first bytes of the connection
+     * @param endPoint The connection EndPoint object
+     * @see #seemsTLS(ByteBuffer)
+     */
+    protected void unknownProtocol(ByteBuffer buffer, EndPoint endPoint)
+    {
+        endPoint.close();
+    }
+
+    private class PlainOrSslConnection extends AbstractConnection implements Connection.UpgradeFrom
+    {
+        private final Connector connector;
+        private final ByteBuffer buffer;
+
+        public PlainOrSslConnection(EndPoint endPoint, Connector connector)
+        {
+            super(endPoint, connector.getExecutor());
+            this.connector = connector;
+            this.buffer = BufferUtil.allocateDirect(1536);
+        }
+
+        @Override
+        public void onOpen()
+        {
+            super.onOpen();
+            fillInterested();
+        }
+
+        @Override
+        public void onFillable()
+        {
+            try
+            {
+                int filled = getEndPoint().fill(buffer);
+                if (filled > 0)
+                {
+                    upgrade(buffer);
```

Comment:
This needs to wait until at least 2 bytes have been received.  If only 1 byte is received it needs to remain fill interested.

Reply:
Ah, good point. I'll revert to 1 byte check.